Fixes potential symlink attack. See Debian Bug #509348.
==========================================================================
--- a/noip2.c
+++ b/noip2.c
@@ -1581,10 +1581,13 @@
 ///////////////////////////////////////////////////////////////////////////
 void dump_buffer(int x)
 {
-	int fd;
+	int fd = -1;
 
 	strcpy(tmp_filename, "/tmp/NO-IPXXXXXX");
-	fd = mkstemp(tmp_filename);
+	if((fd = mkstemp(tmp_filename)) == -1){
+		perror("mkstemp");
+		exit(1);
+	}
 	write(fd, buffer, x);
 	close(fd);
 	Msg("Error info saved in %s", tmp_filename);
